cilium: lock GC walks for global CT maps to serialize deletions

borkmann opened the pull request:
We've seen reports where many goroutines walk the CT map concurrently in order to
erase data from it, namely (*Endpoint).scrubIPsInConntrackTableLocked() seems to
be called from i) REST API handler via (*PutEndpointID).ServeHTTP() ->
deleteEndpointQuiet() -> LeaveLocked(), and ii) (*Endpoint).runPreCompilationSteps().
This is suboptimal as iterations through the BPF map could evict keys the other
routines are currently holding and therefore enforcing restarts of the walk until
we hit the stats.MaxEntries limit of the map w/o having finished the regular walk
in DumpReliablyWithCallback().

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/9645)
<!-- Reviewable:end -->

borkmann committed:
cilium: lock GC walks for global CT maps to serialize deletions

We've seen reports where many goroutines walk the CT map concurrently in order to
erase data from it, namely (*Endpoint).scrubIPsInConntrackTableLocked() seems to
be called from i) REST API handler via (*PutEndpointID).ServeHTTP() ->
deleteEndpointQuiet() -> LeaveLocked(), and ii) (*Endpoint).runPreCompilationSteps().
This is suboptimal as iterations through the BPF map could evict keys the other
routines are currently holding and therefore enforcing restarts of the walk until
we hit the stats.MaxEntries limit of the map w/o having finished the regular walk
in DumpReliablyWithCallback().

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

joestringer reviewed (approved):
I feel like we should move towards coalescing GC attempts rather than blocking all the callers for an arbitrary time, but in practice I suspect that this is an improvement on the existing logic.

Do we have an understanding in mind for how long one of these GC dump periods is? Is it in the order of 200ms, meaning that even if we deleted say 10 pods at once, then they all queue to clean up the CT map, the last of the pod deletions should be able to complete within about 2 seconds? If this takes too long, we could run into client API timeouts or similar which may complain if you try to delete too many pods at once.

borkmann committed:
fqdn: remove EP header file sync from MarkDNSCTEntry

Issue is that SyncEndpointHeaderFile() takes the EP lock, and from other
callsites e.g. scrubIPsInConntrackTable() we also take the EP lock.
Latter is before calling into CT GC, and former is while walking it from
a doFiltering() callback hence if both are the same EPs we might run into
a potential deadlock.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
